src/Codec/Xlsx/Parser.hs

```diff
@@ -39,23 +41,22 @@ import           Codec.Xlsx.Types.Internal.CustomProperties  as CustomProperties
 import           Codec.Xlsx.Types.Internal.Relationships     as Relationships
 import           Codec.Xlsx.Types.Internal.SharedStringTable
 
-
 -- | Reads `Xlsx' from raw data (lazy bytestring)
 toXlsx :: L.ByteString -> Xlsx
-toXlsx = either (error . show) id . toXlsxEither
+toXlsx = either (error . show) id . toXlsxOrError
 
 data ParseError = InvalidZipArchive
                 | MissingFile FilePath
                 | InvalidFile FilePath
                 deriving (Show, Eq)
 
 -- | Reads `Xlsx' from raw data (lazy bytestring), failing with Left on parse error
-toXlsxEither :: L.ByteString -> Either ParseError Xlsx
-toXlsxEither bs = do
-  ar <- first (const InvalidZipArchive) $ Zip.toArchiveOrFail bs
+toXlsxOrError :: MonadError ParseError m => L.ByteString -> m Xlsx
```

Comment:
@qrilka Actually I started out with that approach, but ended up chasing the `MonadError` types down into the core. I had thought your suggestion was that it would be more intention-revealing if the core functions used `throwError` etc. rather than working directly with Either. Hence the new `whenXThrow` combinators below instead of `note`, etc.

Are you suggesting that we keep the inner functions in this module specialized to Either? The big advantage I saw of the rewrite I just made is that the user can determine how errors are to be reported, as long as there is a `MonadError` instance.

Sorry, perhaps I'm not clear on exactly what you're suggesting. I haven't worked a ton with `MonadError` so it's quite possible I'm missing an easier way.